### PR TITLE
SER-1822 - audit setup

### PR DESCRIPTION
**Pull request by Vikeln**
Audit functionality created.
An audit event on an entity can be done as show below, called as an application event

Autowire the ApplicationEventPublisher to trigger async events
  ```
  @Autowired
    private ApplicationEventPublisher publisher;
```
create a new audit event and trigger it
  ```
 NewAuditEvent event = new NewAuditEvent(this, 1, "UPDATE", "AppUser", "", "data", currentUser, "SUCCESS");
 publisher.publishEvent(event);
```

### CODE REVIEW

File: src/main/java/org/apache/fineract/api/AuditApiResource.java [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+@Tag(name = "Users API")
+public class AuditApiResource {
+    private AuditService auditService;
+
+    public AuditApiResource(AuditService auditService) {
+        this.auditService = auditService;
+    }
+
+    @GetMapping
+    @Operation(summary = "List Audits", description = "Get a 200 list of audits that match the criteria supplied and sorted by audit id in descending order, and are within the requestors' data scope. Also it supports pagination and sorting\n" + "\n" + "Example Request:\n" + "\n" + "audits\n" + "\n" + "audits?actionName=UPDATE&entityName=AppUser\n")
```
Comment (amydoxym) [inline src/main/java/org/apache/fineract/api/AuditApiResource.java]:
```suggestion
    @Operation(summary = "List Audits", description = "Get a list of audits that match the criteria supplied and sorted by audit id in descending order, and are within the requestors' data scope. Also it supports pagination and sorting\n" + "\n" + "Example Request:\n" + "\n" + "audits\n" + "\n" + "audits?actionName=UPDATE&entityName=AppUser\n")
```

File: src/main/java/org/apache/fineract/audit/data/AuditSearch.java [hunk 2]
```diff
@@ -0,0 +1,18 @@
+package org.apache.fineract.audit.data;
+
+import lombok.AllArgsConstructor;
+import lombok.Getter;
+
+import java.util.Date;
+
+@Getter
+@AllArgsConstructor
+public class AuditSearch {
```
Comment (amydoxym) [inline src/main/java/org/apache/fineract/audit/data/AuditSearch.java:13]: add javadoc

File: src/main/java/org/apache/fineract/audit/data/AuditSource.java [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+
+import javax.persistence.*;
+import java.util.Date;
+
+@Entity
+@Table(name = "m_audit_source")
+@Getter
+@Setter
+@NoArgsConstructor
+public class AuditSource extends AbstractPersistableCustom<Long> {
```
Comment (amydoxym) [inline src/main/java/org/apache/fineract/audit/data/AuditSource.java:21]: add javadoc

File: src/main/java/org/apache/fineract/audit/data/AuditSource.java [hunk 4]
```diff
@@ -0,0 +1,42 @@
...
+    @Column(name = "data_as_json", length = 1000)
+    private String dataAsJson;
+
+    @ManyToOne
+    @JoinColumn(name = "maker_id", nullable = false)
+    private AppUser maker;
+
+    @Column(name = "made_on_date", nullable = false)
+    @Temporal(TemporalType.TIMESTAMP)
+    private Date madeOnDate;
```
Comment (amydoxym) [inline src/main/java/org/apache/fineract/audit/data/AuditSource.java]: please use LocalDate and if possible LocalDateTime to support timezones

 and `@Temporal(TemporalType.TIMESTAMP)` is no longer needed

File: src/main/java/org/apache/fineract/audit/events/NewAuditEvent.java [hunk 5]
```diff
@@ -0,0 +1,33 @@
+package org.apache.fineract.audit.events;
+
+import lombok.Getter;
+import org.apache.fineract.organisation.user.AppUser;
+import org.springframework.context.ApplicationEvent;
+
+@Getter
+public class NewAuditEvent extends ApplicationEvent {
```
Comment (amydoxym) [inline src/main/java/org/apache/fineract/audit/events/NewAuditEvent.java:13]: add javadoc

File: src/main/java/org/apache/fineract/audit/events/AuditEventListener.java [hunk 6]
```diff
@@ -0,0 +1,28 @@
...
+import lombok.extern.slf4j.Slf4j;
+import org.apache.fineract.audit.service.AuditService;
+import org.apache.fineract.audit.data.AuditSource;
+import org.springframework.context.event.EventListener;
+import org.springframework.scheduling.annotation.Async;
+import org.springframework.stereotype.Component;
+
+@Component
+@Slf4j
+public class AuditEventListener {
```
Comment (amydoxym) [inline src/main/java/org/apache/fineract/audit/events/AuditEventListener.java:15]: add javadoc

File: src/main/java/org/apache/fineract/audit/service/AuditServiceImpl.java [hunk 7]
```diff
@@ -0,0 +1,40 @@
...
+@Slf4j
+public class AuditServiceImpl implements AuditService{
+    private final AuditSourceRepository auditSourceRepository;
+
+    public AuditServiceImpl(AuditSourceRepository auditSourceRepository) {
+        this.auditSourceRepository = auditSourceRepository;
+    }
+
+    @Override
+    public AuditSource createNewEntry(NewAuditEvent event) {
```
Comment (amydoxym) [inline src/main/java/org/apache/fineract/audit/service/AuditServiceImpl.java:25]: please add @Transactional

File: src/main/java/org/apache/fineract/audit/service/AuditServiceImpl.java [hunk 8]
```diff
@@ -0,0 +1,40 @@
...
+        audit.setActionName(event.getActionName());
+        audit.setDataAsJson(event.getDataAsJson());
+        audit.setProcessingResult(event.getProcessingResult());
+        audit.setMadeOnDate(new Date());
+        audit.setMaker(event.getMaker());
+        return auditSourceRepository.save(audit);
+    }
+
+    @Override
+    public Page<AuditSource> getAudits(Specification<AuditSource> specification, Pageable pageable) {
```
Comment (amydoxym) [inline src/main/java/org/apache/fineract/audit/service/AuditServiceImpl.java:39]: Add @Transactional(readOnly=true)

**Review by amydoxym — CHANGES_REQUESTED** [review]

**Review by victorjbassey — APPROVED** [review]

**Review by gilclefes — APPROVED** [review]
looks good to me.

**Review by amydoxym — APPROVED** [review]

**Pull request merged**